### PR TITLE
sql/schemachanger/scplan: allow plan to move to NonRevertible earlier

### PR DESCRIPTION
**Pull request by ajwerner**
This is critical for DROP COLUMN. In `DROP COLUMN` we cannot perform the
primary index swap until the dropping column reaches `DELETE_ONLY`, which is
not revertible. The primary index swap itself is revertible. Given this fact,
we need a mechanism to be able to "promote" revertible operations (operations
which don't destroy information or cause irrevocable visible side effects) to
be grouped with or after non-revertible operations. This commit makes that
happen naturally.

Release note: None

### CODE REVIEW

File: pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go [hunk 1]
```diff
@@ -695,6 +695,47 @@ func init() {
...
+				//
+				// TODO(ajwerner): It'd be nice to be able to express this in rel
+				// directly.
+				rel.Filter("indexes-id-less", "a", "b")(func(a, b scpb.Element) bool {
+					aID, _ := screl.GetIndexID(a)
+					bID, _ := screl.GetIndexID(b)
+					return aID < bID
+				}),
+			}
+		})
```
Comment (postamar) [inline pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go:738]: FWIW I'm totally fine with these, they are well-motivated and easy to understand.

File: pkg/sql/schemachanger/scplan/internal/scstage/build.go [hunk 2]
```diff
@@ -563,6 +628,26 @@ func (bc buildContext) nodes(current []scpb.Status) []*screl.Node {
...
+		if !ok {
+			return false
+		}
+		return rec(oe.To()) || oe.Type() != scop.MutationType
+	}
+	for _, n := range nodes(g, init.Targets, init.Current) {
+		rec(n)
+	}
+	return m
+}
```
Comment (postamar) [inline pkg/sql/schemachanger/scplan/internal/scstage/build.go:649]: Ah. This makes sense. 

This logic belongs in the bowels of opgen, though, along with what determines the `revertible` boolean field on each op-edge. You'll find that it's quite similar. 

In fact, perhaps this `revertible` boolean field should be replaced by an enum with three states:
- revertible, failure possible;
- revertible, failure not possible;
- not revertible, failure not possible.

**Review by postamar — APPROVED** [review]
LGTM but please that can-still-fail determining logic to opgen as suggested, it should be quite straightforward to do.

File: pkg/sql/schemachanger/scplan/internal/scstage/build.go (hunk 2 above)
Comment (ajwerner) [reply pkg/sql/schemachanger/scplan/internal/scstage/build.go]: I moved this into opgen in the newly added first commit. It was the right thing to do, but took a little bit of elbow grease. 

I've left two booleans as I found it easier to deal with for now, do you feel strongly?

Comment (postamar) [reply pkg/sql/schemachanger/scplan/internal/scstage/build.go]: No, I don't feel strongly about that.

**Review by postamar — APPROVED** [review]

**Pull request merged**